src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -62,15 +67,342 @@ int
+    check_empty();
+    return m_axlf;
+  }
+
+};
+
+} //namespace
+
+////////////////////////////////////////////////////////////////
+// xrt_xclbin C++ API implementations (xrt_xclbin.h)
+////////////////////////////////////////////////////////////////
+namespace xrt {
+
+xclbin::
+xclbin(const std::string &filename) : handle(std::make_shared<xclbin_impl>(filename))
+{}
+
+xclbin::
+xclbin(const std::vector<char> &data) : handle(std::make_shared<xclbin_impl>(data))
+{}
+
+std::vector<std::string>
+xclbin::
+get_cu_names() const
+{
+  return this->get_handle()->get_cu_names();
+}
+
+std::string
+xclbin::
+get_xsa_name() const
+{
+  return this->get_handle()->get_xsa_name();
+}
+
+uuid
+xclbin::
+get_uuid() const
+{
+  return this->get_handle()->get_uuid();
+}
+
+const std::vector<char> &
+xclbin::
+get_data() const
+{
+  return this->get_handle()->get_data();
+}
+
+}
+
+////////////////////////////////////////////////////////////////
+// xrt_xclbin API implementations (xrt_xclbin.h)
+////////////////////////////////////////////////////////////////
+
+// C-API handles that must be explicitly freed. Corresponding managed
+// handles are inserted in this map.  When the unmanaged handle is
+// freed, it is removed from this map and underlying object is
+// deleted if no other shared ptrs exists for this xclbin object
+static std::map<xrtXclbinHandle, std::shared_ptr<xrt::xclbin_impl>> xclbins;
+
+static std::shared_ptr<xrt::xclbin_impl>
+get_xclbin(xrtXclbinHandle handle)
+{
+  auto itr = xclbins.find(handle);
+  if (itr == xclbins.end())
+    throw xrt_core::error(-EINVAL, "No such xclbin handle");
+  return (*itr).second;
+}
+
+static void
+free_xclbin(xrtXclbinHandle handle)
+{
+  if (xclbins.erase(handle) == 0)
+    throw xrt_core::error(-EINVAL, "No such xclbin handle");
+}
+
+// Utility function for device class to verify that the C xclbin handle is valid
+// Needed when the C API for device tries to load an xclbin using C pointer to xclbin
+namespace xrt {
+namespace xclbin_int {
+bool
+is_valid(xrtXclbinHandle handle)
+{
+  bool rtn = (xclbins.find(handle) != xclbins.end());
+  return rtn;
+}
+}
+};
+
+xrtXclbinHandle
+xrtXclbinAllocFilename(const char *filename)
+{
+  try {
+    auto xclbin = std::make_shared<xrt::xclbin_impl>(filename);
+    auto handle = xclbin.get();
+    xclbins.emplace(std::make_pair(handle, std::move(xclbin)));
+    return handle;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    errno = ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    errno = 0;
+  }
+  return nullptr;
+}
+
+xrtXclbinHandle
+xrtXclbinAllocRawData(const char *data, const int size)
+{
+  try {
+    char *data_c = const_cast<char *>(data);
+    std::vector<char> raw_data(data_c, data_c + size);
+    auto xclbin = std::make_shared<xrt::xclbin_impl>(raw_data);
+    auto handle = xclbin.get();
+    xclbins.emplace(std::make_pair(handle, std::move(xclbin)));
```

Comment:
If you are using `emplace()`, could it be just
```c++
     xclbins.emplace(handle, std::move(xclbin));
```
https://en.cppreference.com/w/cpp/container/map/emplace